chain/sub/bcast/consistent.go

```diff
@@ -0,0 +1,180 @@
+	if !ok {
+		return 0
+	}
+	return len(v.(*blksInfo).blks)
+}
+
+type ConsistentBCast struct {
+	lk    sync.RWMutex
+	delay time.Duration
+	// FIXME: Make this a slice??? Less storage but needs indexing logic.
+	m map[abi.ChainEpoch]*bcastDict
+}
+
+func newBcastDict() *bcastDict {
+	return &bcastDict{new(sync.Map)}
+}
+
+// TODO: the VRFProof may already be small enough so we may not need to use a hash here.
+// we can maybe bypass the useless computation.
+func BCastKey(bh *types.BlockHeader) (multihash.Multihash, error) {
+	k := make([]byte, len(bh.Ticket.VRFProof))
+	copy(k, bh.Ticket.VRFProof)
+	binary.PutVarint(k, int64(bh.Height))
+	return multihash.Sum(k, multihash.SHA2_256, -1)
+}
+
+func NewConsistentBCast(delay time.Duration) *ConsistentBCast {
+	return &ConsistentBCast{
+		delay: delay,
+		m:     make(map[abi.ChainEpoch]*bcastDict),
+	}
+}
+
+func cidExists(cids []cid.Cid, c cid.Cid) bool {
+	for _, v := range cids {
+		if v == c {
+			return true
+		}
+	}
+	return false
+}
+
+func (bInfo *blksInfo) eqErr() error {
+	bInfo.cancel()
+	return fmt.Errorf("different blocks with the same ticket already seen")
+}
+
+func (cb *ConsistentBCast) Len() int {
+	cb.lk.RLock()
+	defer cb.lk.RUnlock()
+	return len(cb.m)
+}
+
+func (cb *ConsistentBCast) RcvBlock(ctx context.Context, blk *types.BlockMsg) {
+	cb.lk.Lock()
+	bcastDict, ok := cb.m[blk.Header.Height]
+	if !ok {
+		bcastDict = newBcastDict()
+		cb.m[blk.Header.Height] = bcastDict
+	}
+	cb.lk.Unlock()
+	key, err := BCastKey(blk.Header)
+	if err != nil {
+		log.Errorf("couldn't hash blk info for height %d: %s", blk.Header.Height, err)
+		return
+	}
+	blkCid := blk.Cid()
+
+	bInfo, ok := bcastDict.load(key)
+	if ok {
+		if len(bInfo.blks) > 1 {
+			log.Errorf("equivocation detected for height %d: %s", blk.Header.Height, bInfo.eqErr())
+			return
+		}
+
+		if !cidExists(bInfo.blks, blkCid) {
+			bcastDict.store(key, &blksInfo{bInfo.ctx, bInfo.cancel, append(bInfo.blks, blkCid)})
+			log.Errorf("equivocation detected for height %d: %s", blk.Header.Height, bInfo.eqErr())
+			return
+		}
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, cb.delay)
+	bcastDict.store(key, &blksInfo{ctx, cancel, []cid.Cid{blkCid}})
+}
+
+func (cb *ConsistentBCast) WaitForDelivery(bh *types.BlockHeader) error {
+	cb.lk.RLock()
+	bcastDict := cb.m[bh.Height]
+	cb.lk.RUnlock()
+	key, err := BCastKey(bh)
+	if err != nil {
+		return err
+	}
+	bInfo, ok := bcastDict.load(key)
+	if !ok {
+		return fmt.Errorf("something went wrong, unknown block with Epoch + VRFProof (cid=%s) in consistent broadcast storage", key)
+	}
+	// Wait for the timeout
+	<-bInfo.ctx.Done()
+	if bcastDict.blkLen(key) > 1 {
+		return fmt.Errorf("equivocation detected for epoch %d. Two blocks being broadcast with same VRFProof", bh.Height)
+	}
+	return nil
+}
+
+func (cb *ConsistentBCast) GarbageCollect(currEpoch abi.ChainEpoch) {
+	cb.lk.Lock()
+	defer cb.lk.Unlock()
+
+	// keep currEpoch-2 and delete a few more in the past
```

Comment:
I'd consider doing a "fuller" (eg. lookback the past 24 hours) GC every (say) 1000 epochs or so. It should be pretty fast, and might be necessary to catch any escapees ;)